// rettiwt-core/dist/enums/Analytic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMetric = exports.EGranularity = void 0;
/**
 * Granularity types for analytics.
 *
 * @public
 */
var EGranularity;
(function (EGranularity) {
    EGranularity["DAILY"] = "Daily";
    EGranularity["WEEKLY"] = "Weekly";
    EGranularity["MONTHLY"] = "Monthly";
})(EGranularity = exports.EGranularity || (exports.EGranularity = {}));
/**
 * The different types of metric someone can request.
 *
 * @public
 */
var EMetric;
(function (EMetric) {
    EMetric["ENGAGEMENTS"] = "Engagements";
    EMetric["IMPRESSIONS"] = "Impressions";
    EMetric["PROFILE_VISITS"] = "ProfileVisits";
    EMetric["FOLLOWS"] = "Follows";
    EMetric["VIDEO_VIEWS"] = "VideoViews";
    EMetric["REPLIES"] = "Replies";
    EMetric["LIKES"] = "Likes";
    EMetric["RETWEETS"] = "Retweets";
    EMetric["MEDIA_VIEWS"] = "MediaViews";
    EMetric["BOOKMARK"] = "Bookmark";
    EMetric["SHARE"] = "Share";
})(EMetric = exports.EMetric || (exports.EMetric = {}));
